### PR TITLE
Delay WASM Branch Condition Concretization

### PR DESCRIPTION
**Pull request by jimpo**
The Wasm platform can sometimes silently skip execution paths due to the way it concretizes branch conditions. The issue is that Wasm branch instructions take integers as input and the Wasm symbolic execution engine concretizes the whole integer to continue execution, instead of just the binary condition needed. Aside from being inefficient, the concretization logic prevents forking into more than a small number of values and silently skips exploration of other values.

This change proposes a way to concretize boolean conditions used as advice for execution, imposing the minimal constraints necessary. The engine raises a ConcretizeCondition exception to determine the branch path and constrain the state minimally. The feasible boolean values are put on the stack and read when the instruction in re-executed. If there's a better approach, please do let me know.

### CODE REVIEW

File: manticore/wasm/structure.py [hunk 1]
```diff
@@ -1900,3 +1956,18 @@ def allocate(
...
+    """Tells Manticore to concretize a condition required to direct execution."""
+
+    def __init__(self, message: str, condition: Bool, **kwargs):
+        """
+        :param message: Debug message describing the reason for concretization
+        :param condition: The boolean expression to concretize
+        """
+
+        def setstate(state, value: bool):
+            state.platform.stack.data[-1] = Condition(value)
```
[the remark this answers is not in the document]
Comment (jimpo) [reply manticore/wasm/structure.py:1971]: Right. I tried that first, but `StateBase.__enter__` does not make a copy of the execution state stored on the platform, namely the stack. I'm still not familiar enough with the architecture to know how to fix that one properly...

File: manticore/wasm/structure.py [hunk 2]
```diff
@@ -1667,16 +1707,31 @@ def __init__(self, arity, frame, expected_block_depth=0):
         self.expected_block_depth = expected_block_depth
 
 
+@dataclass
+class Condition:
```
[the remark this answers is not in the document]
Comment (jimpo) [reply manticore/wasm/structure.py:1711]: My intention in defining a new is just for semantics. Having a raw boolean on the stack would be kind of confusing and out of context.

**Comment by jimpo** [comment]
@ehennenfent Thanks for giving this a look!

> Finally, as a structural suggestion - when building this module, I generally tried to avoid putting anything on the stack that wasn't explicitly specified by the WASM spec. That's because stack operations are (thanks to the events API) somewhat user-facing, and I wanted to avoid any undue confusion that would result from having Manticore do internal record-keeping on the stack. Based on my reading of this, it seems like there should never be more than one Condition on the stack at any given time, and at that, it should always be on top. With that in mind, could we instead add a branch_condition field to the ModuleInstance and store the Condition there?

I was intending that there could be multiple Conditions on the stack as a solution to #1642 for Wasm. In particular, on an `i32.div_s` instruction, the engine would concretize both the zero-division and division-overflow conditions, ultimately getting both Conditions pushed to the top of the stack and proceeding accordingly. So I think some sort of stack is the right abstraction. But I understand the desire to use a separate structure. I view the Conditions as "execution advice". Perhaps it makes sense to have another stack on `WasmWorld`, like `condition_stack` or `aux_stack` or `advice_stack` or something?

File: manticore/wasm/structure.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ehennenfent) [reply manticore/wasm/structure.py:1711]: > Having a raw boolean on the stack would be kind of confusing and out of context
Agree with this, but I think the `Condition` class and the `BoolConstant` class serve more or less the same purpose. If you look at the implementations, they're almost identical

```
class Condition:
    value: bool  #: The boolean condition value

    def __init__(self, value: bool):
        self.value = value
```

```
class BoolConstant(Bool):
    __slots__ = ["_value"]

    def __init__(self, value: bool, *args, **kwargs):
        self._value = value
        super().__init__(*args, **kwargs)

    def __bool__(self):
        return self.value

    @property
    def value(self):
        return self._value
```

**Comment by ehennenfent** [comment]
>  Perhaps it makes sense to have another stack on `WasmWorld`, like `condition_stack` or `aux_stack` or `advice_stack` or something?
This sounds good to me! We actually already use one auxiliary stack for control flow (`block_depths` in `ModuleInstance`) so that we can correctly flatten the nested expression blocks into a single instruction queue.

**Comment by ehennenfent** [comment]
Looks like Github Actions' checkout@v1 script is having some issues right now. If you update from `master`, it should bump it to checkout@v2.

**Comment by jimpo** [comment]
@ehennenfent Yes, in part I pass the advice to `ConcretizeCondition` to circumvent the bug with state copying in `ManticoreBase._fork`. But the bigger motivation is that I think it's easiest to clear `advice` on the state after each instruction execution, regardless of outcome. Another behavior that could work is to clear the advice on any outcome of instruction execution aside from a `Concretize` exception (which is the only case where the instruction execution didn't complete/resolve and the last instruction is requeued).

And yes, the thought with allowing for multiple conditions in the advice is to handle `i32.div_s`.

**Comment by ehennenfent** [comment]
Glad we're on the same page. I do find all the shuffling-around of `advice` to be a bit convoluted, but we can fix that when we get around to fixing state copying. Thanks again for the PR!

**Pull request merged**